### ISSUE TITLE
Question: hierarchical configuration

### ISSUE THREAD

**Issue by zexuan-zhou** [dynaconf/dynaconf#644]
Hi team,

I'm new to this library and it looks pretty cool. I'd like to give it a try but I wonder if this library support hierarchical configuration setup. That is, I have some global (or project-wise) configuration and I have some file-related (or submodule-specific) configuration. I wonder if this library can detect such pattern and build a configuration object out of it.

An example of hierarchical configuration I found is here: https://github.com/adobe/himl#deep-merge-example
But it seems that they only support yaml.

Thank you

**Comment by rochacbruno** [comment]
Hi @zexuan-zhou 

Yes, there are multiple ways to achieve this.

Most explained on https://www.dynaconf.com/merging/

some here: (will add to the docs in future)


## Single file file with merge_enabled

settings.yaml
```yaml
default:
  deep:
    key1: v1
    key2: v2
  deep_list:
    - item1
    - item2

prod:
  deep:
    key3: v3
  deep_list:
    - item3
```

app.py
```py
from dynaconf import Dynaconf
settings = Dynaconf(
	settings_file="settings.yaml",
	merge_enabled=True,
	environments=True,
	envvar_switcher="ENV",
)

print(settings.deep)
print(settings.deep_list)
```

```bash
ENV=prod python app.py
{"key1": "v1", "key2": "v2", "key3": "v3"}
["item1", "item2", "item3"]
```

## Single settings file with local merge

settings.yaml
```yaml
default:
  deep:
    key1: v1
    key2: v2
  deep_list:
    - item1
    - item2

prod:
  dynaconf_merge: true
  deep:
    key3: v3
  deep_list:
    - item3
```

app.py
```py
from dynaconf import Dynaconf
settings = Dynaconf(
	settings_file="settings.yaml",
	merge_enabled=False,  # default to False 
	environments=True,
	envvar_switcher="ENV",
)

print(settings.deep)
print(settings.deep_list)
```

```bash
ENV=prod python app.py
{"key1": "v1", "key2": "v2", "key3": "v3"}
["item1", "item2", "item3"]
```

## Single settings file with specific merge

settings.yaml
```yaml
default:
  deep:
    key1: v1
    key2: v2
  deep_list:
    - item1
    - item2

prod:
  deep:
    key3: v3
    dynaconf_merge: true
  deep_list:
    - item3
    - dynaconf_merge
```

app.py
```py
from dynaconf import Dynaconf
settings = Dynaconf(
	settings_file="settings.yaml",
	merge_enabled=False,  # default to False 
	environments=True,
	envvar_switcher="ENV",
)

print(settings.deep)
print(settings.deep_list)
```

```bash
ENV=prod python app.py
{"key1": "v1", "key2": "v2", "key3": "v3"}
["item1", "item2", "item3"]
```

## Multiple settings files

settings1.yaml
```yaml
deep:
  key1: v1
  key2: v2
deep_list:
  - item1
  - item2
```

settings2.yaml
```yaml
deep:
  key3: v3
deep_list:
  - item3
```

app.py
```py
from dynaconf import Dynaconf
settings = Dynaconf(
	settings_files=["settings1.yaml", "settings2.yaml"],
  	merge_enabled=True,  # enable merging, can also be done specifically
	environments=False,  # default to False
	envvar_switcher="ENV",
)

print(settings.deep)
print(settings.deep_list)
```

```bash
ENV=prod python app.py
{"key1": "v1", "key2": "v2", "key3": "v3"}
["item1", "item2", "item3"]
```


## Pre load

settings_default.yaml
```yaml
deep:
  key1: v1
  key2: v2
deep_list:
  - item1
  - item2
```

settings.yaml
```yaml
deep:
  key3: v3
deep_list:
  - item3
```

settings.local.yaml
```yaml
deep:
  key4: v4
deep_list:
  - item4
```

app.py
```py
from dynaconf import Dynaconf
settings = Dynaconf(
    preload=["settings_default.yaml"],
	settings_files=["settings.yaml"],  # will also load `settings.local.yaml`
  	merge_enabled=True,  # enable merging, can also be done specifically
	environments=False,  # default to False
	envvar_switcher="ENV",
)

print(settings.deep)
print(settings.deep_list)
```

```bash
ENV=prod python app.py
{"key1": "v1", "key2": "v2", "key3": "v3", "key4": "v4"}
["item1", "item2", "item3", "item4"]
```

## Manual loading

settings.default.yaml
```yaml
deep:
  key1: v1
  key2: v2
deep_list:
  - item1
  - item2
```

settings.yaml
```yaml
deep:
  key3: v3
deep_list:
  - item3
```

settings.other.yaml
```yaml
deep:
  key4: v4
deep_list:
  - item4
```

app.py
```py
from dynaconf import Dynaconf
settings = Dynaconf(
  	merge_enabled=True,  # enable merging, can also be done specifically
	environments=False,  # default to False
	envvar_switcher="ENV",
)

# No settings_files was defined, lets load it in order 

settings.load_file("settings.default.yaml")
settings.load_file("settings.yaml")
settings.load_file("settings.other.yaml")

print(settings.deep)
print(settings.deep_list)
```

```bash
ENV=prod python app.py
{"key1": "v1", "key2": "v2", "key3": "v3", "key4": "v4"}
["item1", "item2", "item3", "item4"]
```

## Envvar merging 

```bash
export DYNACONF_DEEP__KEY5=5 
export DYNACONF_DEEP="@merge key5=5" 
export DYNACONF_DEEP_LIST="@merge item5"
```

**Comment by zexuan-zhou** [comment]
Thank you. This looks great!

**Comment by zexuan-zhou** [comment]
@rochacbruno I tried to use dynaconf with the following script but it doesn't seem to work as I expected
```
# default_config/__init__.py
from dynaconf import Dynaconf

settings = Dynaconf(
    envvar_prefix="default",
    environments=True,
    setting_files=['settings.py']
)

print(settings.PYTORCH)

# default_config/settings.py
PYTORCH = "1.7.1"

# command 
python default_config/__init__.py 
```
it gives me `AttributeError: 'Settings' object has no attribute 'PYTORCH'`
Any idea?

**Comment by rochacbruno** [comment]
`❯ cat default_config/__init__.py`
```py
from dynaconf import Dynaconf

settings = Dynaconf(
    environments=True,
    settings_files=['settings.py'],
)

print(settings.PYTORCH)
```

`❯ cat default_config/settings.py` 
```py
PYTORCH = "1.7.1"
```


`❯ python default_config/__init__.py`
```bash
1.7.1
```

@zexuan-zhou I just tried and worked

**Comment by zexuan-zhou** [comment]
@rochacbruno Thank you. I deleted all files and recreated them and now it works.
I wonder if there is a way to print all configurations of a `Dynaconf` object?
`print(settings)` simply outputs something like this `<dynaconf.base.Settings object at 0x7fd0cc852dd0>`

**Comment by rochacbruno** [comment]
@zexuan-zhou https://www.dynaconf.com/advanced/#exporting

`settings.as_dict()`

**Comment by zexuan-zhou** [comment]
> https://www.dynaconf.com/advanced/#exporting

@rochacbruno When I try to predict the settings output by `settings.as_dict()`, it doesn't seem to do a good job on loading all nested values as nested dictionary. Some of the values are a string of a dictionary. Is it possible to pretty print the setting object? Thank you

**Comment by rochacbruno** [comment]
@zexuan-zhou can you provide an example?

for me it works like this

```py
In [2]: settings.as_dict()
Out[2]: 
{'LOAD_DOTENV': True,
 'DEFAULT_SETTINGS_PATHS': ['settings.py',
  'settings.toml',
  'settings.tml',
  'settings.yaml',
  'settings.yml',
  'settings.ini',
  'settings.conf',
  'settings.properties',
  'settings.json',
  '.secrets.py',
  '.secrets.toml',
  '.secrets.tml',
  '.secrets.yaml',
  '.secrets.yml',
  '.secrets.ini',
  '.secrets.conf',
  '.secrets.properties',
  '.secrets.json'],
 'A_BIG_DICT': {'file': {'.env': True,
   'settings.json.global': True,
   'settings.json.default': True,
   'settings.properties': 'True',
   'settings.conf': 'True',
   'settings.ini': 'True',
   'settings.yml': True,
   'settings.yaml': True,
   'settings.tml.global': True,
   'settings.tml.default': True,
   'settings.toml.global': True,
   'settings.toml.default': True,
   'global_settings.py': True,
   'development_settings.py': True,
   'settings.py': True},
  'nested_1': {'nested_2': {'nested_3': {'nested_4': {'json': True,
      'yaml': True,
      'toml': True,
      'py': True}}}},
  'never_erased': True},
 'PYTHON_VAR': 1,
 'PYTHON_DEV_VAR': 1,
 'PYTHON_GLOBAL_VAR': 1,
 'TOML_VAR': 1,
 'TOML_GLOBAL_VAR': 3,
 'TML_VAR': 1,
 'TML_GLOBAL_VAR': 3,
 'YAML_VAR': 1,
 'YML_VAR': 1,
 'INI_VAR': '1',
 'CONF_VAR': '1',
 'PROPERTIES_VAR': '1',
 'JSON_VAR': 1,
 'ENV_VAR': 1,
 'WORKS': 'merge_configs'}

```

**Comment by zexuan-zhou** [comment]
@rochacbruno Here is an example

`cat default_config/__init__.py `
```
from dynaconf import Dynaconf
import json

default_config = Dynaconf(
    settings_files=[
        'config.yaml',  # default config
    ],
)

print(f"default_config: {json.dumps(default_config.as_dict(), indent=2)}")
```

`cat default_config/config.yaml `
```
A:
  a: a_value
  b:
    bb: bb_value
    ba: "@format {this.A.a}"
  c:
    cc: "@format {this.A.b}"
B:
  a:
    a1:
      - ["a11", "a12"]
      - ["a13", "a14"]
  inference_cols: "@jinja {{this.B.a.a1|map(attribute=0)|list}}"
```
Run it with `python default_config/__init__.py ` and I got
```
default_config: {
  "A": {
    "a": "a_value",
    "b": {
      "bb": "bb_value",
      "ba": "a_value"
    },
    "c": {
      "cc": "{'bb': 'bb_value', 'ba': 'a_value'}"  ## this should be a dictionary but it's printed as a string
    }
  },
  "B": {
    "a": {
      "a1": [
        [
          "a11",
          "a12"
        ],
        [
          "a13",
          "a14"
        ]
      ]
    },
    "inference_cols": "['a11', 'a13']"  ## I don't know why my jinja expression gives this as a string instead of an array
  }
}
```
Thank you

**Comment by rochacbruno** [comment]
@zexuan-zhou this is related to another issue #642

**Comment by jgod** [comment]
Is there a way to disable deep merging lists (instead, override and just keep latest list), but keeping the behavior for objects?

**Comment by pedro-psb** [comment]
> Is there a way to disable deep merging lists (instead, override and just keep latest list), but keeping the behavior for objects?

@jgod I believe there is not. Maybe this could be implemented as extended options for `merge_enabled`.

```python
merge_enabled = "list" # merge only list
merge_enabled = "dict" # merge only dicts/objects
merge_enabled = True or "all" # merge all
merge_enabled = False # merge nothing
```

**Comment by pedro-psb** [comment]
> I tried to put dynaconf_merge_unique as a key in the yaml file under deep_list but it doesn't seem to dedup for me. 

@zexuan-zhou Currently (`v3.1.12`) this works as described in [the mentioned section](https://www.dynaconf.com/merging/#avoiding-duplications-on-lists) of the docs. Although there is not a `yaml` example there, the principle is the same.

```yaml
# cat config.yaml 
deep:
  key1: v1
deep_list:
  - item1
  - item2

# cat overwrite_config.yaml 
deep:
  key1: v3
  key2: v2
deep_list:
  - item1  # I want to only keep item1 for deep_list
  - dynaconf_merge_unique
```

```py
>>> settings.deep_list
['item2', 'item1']
```

**Comment by pedro-psb** [comment]
> `cuda_visible_devices: "@jinja {% for i in range(0, this.training.gpu_resource_config.n_gpus) %} {{i}} {% endfor %} | join(',')"`
> gives me `"cuda_visible_devices": " 0  1  2  3  4  5  6  7  | join(',')"`. I want the output to be something like `"0,1,2,3,4,5,6,7"`.
> How can I use a jinja2 expression with filter in the same interpolation expression?

@zexuan-zhou I guess it is not possibly to do this exatcly that way, because *filters* in jinja2 must be used inside *expressions*, like `{{ expr | filter }}`. Currently, if it really required to do this kind of processing inside the templates, you could do something like that:

```yaml
gpu_no: 8
_cuda_visible_devices: "@jinja {% for i in range(0, this.gpu_no) %}{{i}}{% endfor %}"
cuda_visible_devices: "@jinja {{ this._cuda_visible_devices | join(',') }}"
```
```py
>>> settings._cuda_visible_devices
"01234567"

>>> settings.cuda_visible_devices
"0,1,2,3,4,5,6,7"
# join(',') would cause problems for 2-digit numbers, 
# because it appears to join each character of the string.
```
---

But, in this case, I think this would be a better approach:

```yaml
gpu_no: 8
cuda_visible_devices: "@jinja {% for i in range(0, this.gpu_no) %}{{i}}{% if not loop.last %},{% endif %}{% endfor %}"

# or cast to a list
cuda_visible_devices_list: "@json @jinja [{% for i in range(0, this.gpu_no) %}{{i}}{% if not loop.last %},{% endif %}{% endfor %}]"
```
```py
>>> settings.cuda_visible_devices
"0,1,2,3,4,5,6,7"

>>> settings.cuda_visible_devices_list
[0, 1, 2, 3, 4, 5, 6, 7]
```

**Comment by pedro-psb** [comment]
I think the main doc change needed here is adding [these examples](https://github.com/dynaconf/dynaconf/issues/644#issuecomment-902313800) about merging.

I've reviewed the [Merging](https://www.dynaconf.com/merging/) page to see where I could fit those examples, and they could go in the *More Examples*, which have a link to some code in the repo (currently a broken link). Still, I've concluded it would be best to *restructure the page*, similarly to what I've proposed in #911. Basically, I'd propose a succinct Usage section and a more comprehensive Examples section (where this could go).

There are other implicit doc improvement suggested by this thread. The one I think would be great is to have a dedicated section about available tokens (`@tokens`) and their roles (the specific comment here was suggesting just to list all available type castings). This section would be a good place to document the new `add_converter` utility too, but I'm not sure where it would fit yet.

If this is ok, I'll close the issue and split it in two so it is easier to track what is pending.
